zed-editor: respect user interactivity with settings and keymaps

yajo opened the pull request:
### Description
Merges frozen config from HM into dirty config.

Fixes https://github.com/nix-community/home-manager/issues/6835
<!--

Please provide a brief description of your change.

-->

### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://nix-community.github.io/home-manager/#sec-guidelines

-->

- [x] Change is backwards compatible.

- [x] Code formatted with `nix fmt` or `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`
    or `nix build --reference-lock-file flake.lock ./tests#test-all` using Flakes.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

#### Maintainer CC

<!--
If you are updating a module, please @ people who are in its `meta.maintainers` list.
If in doubt, check `git blame` for whoever last touched something.
-->
@libewa

khaneliman commented:
rebase to resolve CI

yajo commented:
Done. I also modified the script a little bit to ease automatic migrations from older versions of the module.

khaneliman commented:
```bash
mkdir: cannot create directory '/build': Read-only file system
building '/nix/store/jwhk0sd3ny8dl0dm29yjmmia96ny88sw-nmt-test-zed-extensions.drv'...
error: builder for '/nix/store/arl6x2ra1fqwvbjjpnsrs8yiprsqcrjz-nmt-test-zed-keymap.drv' failed with exit code 1;
       last 1 log lines:
       > mkdir: cannot create directory '/build': Read-only file system
       For full logs, run:
         nix log /nix/store/arl6x2ra1fqwvbjjpnsrs8yiprsqcrjz-nmt-test-zed-keymap.drv
```

yajo commented on tests/modules/programs/zed-editor/extensions.nix:
```diff
@@ -11,20 +11,50 @@
     ];
   };
 
+  home.homeDirectory = lib.mkForce "/build/hm-user";
```
It looks like MacOS doesn't use `/build/` as the directory where builds are performed. I only have Linux machines. Do you have any idea on what's the correct value here for Mac so I can fix the test?

khaneliman replied:
I have a macbook. I can take a look at this later, when I get time.

khaneliman replied:
Sorry, i forgot about this...

nickdichev commented:
I was able to build this on aarch64-darwin with your latest commit

khaneliman commented:
Looks like formatting failed.

yajo commented:
Thanks, I reformatted.

yajo commented:
So, could this one be merged?

libewa commented:
Other editors (code) also have something like `mutableSettings`

khaneliman commented:
> Other editors (code) also have something like `mutableSettings`

I see `mutableExtensionsDir` but not `mutableSettings`. Right now, the home manager module, from my understanding, has a similar limitation to this zed-editor one. Where we write a read only file and vscode tries to update it and fails.

yajo commented:
Immutable settings in zed are extremely uncomfortable. Many of its buttons only work when writing to its settings file. Zed is not designed to work without write access to its settings: https://github.com/zed-industries/zed/issues/24537#issuecomment-2784496147

I understand your concerns, but I'm not sure implementing immutable settings has any use for Zed.

khaneliman commented:
@libewa any opinion on implementation?

yajo commented:
After trying it locally, I have changed a bit the algorithm.

I removed the part that was checking for a read-only file. Because of how the DAG is set, those symlinks will be removed automatically when we get to this part of the script. Thus, there's no need to check that.

I also added an argument to set the default value for an empty file. `jq` wasn't able to multiply an empty file with an object.

nyabinary commented:
```error: 'thefuck' has been removed due to lack of maintenance upstream and incompatible with python 3.12+. Consider using 'pay-respects' instead```

nyabinary commented:
@khaneliman should be good now with that recent push

Masrepus commented:
I just noticed a small issue this PR introduced: In case you don't have an existing keymap.json file and you also didn't set `userKeymaps`, the default value provided by the Nix module is going to be `{}` instead of `[]`. This leads to some jq issues while activating the config and ultimately to Zed complaining about an invalid keymap file. Manually setting it to `[]` resolves it, so I believe the default should be updated

andresilva commented:
https://github.com/nix-community/home-manager/pull/7222

CodedNil commented:
Is there any way to opt out of this change? I have noticed that if I change settings for example adding experimental theme overrides, any time i remove an option in the declarative it will just linger and i have to delete the settings file. I personally would never use any of the zed settings ui imperatively.

yajo commented:
Currently there's no opt-out option because of https://github.com/nix-community/home-manager/pull/6993#issuecomment-2898561400.

However if you still have use for it, you can open a PR 😊. A very simple implementation would be to delete the preexisting files before running current script. Since it would merge with nothing, it would just restore your nix stuff always.

woile commented:
This feature has helped me with the migration of the settings, as the editor was reporting deprecated settings and thanks to this I was able to go back to nix and update them easily. It would be nice to have a setting to be able to switch between mutable and immutable, but if there's no work on it, I'd rather have it like it's now.

yajo commented:
Zed is designed to have writable settings files. Like it or not, that's how it's done. [Upstream has confirmed that's their vision on this subject](https://github.com/zed-industries/zed/issues/24537#issuecomment-2784496147).

Those files [can even contain secrets](https://github.com/akbxr/gitlab-mcp-zed?tab=readme-ov-file#required-settings), which you don't want in your Nix store.

Just the work of adding a simple hotkey was a nightmare with the previous implementation. You had to:
1. Copy the `keymaps.json` file contents.
2. Delete it because it was a symlink to the nix store.
3. Create a new one.
4. Paste the contents.
6. Change the keymap.
8. Test it works.
9. Copy what you've done into your nix configuration, converted into Nix.
11. Remove your `keymaps.json` file again.
12. Re-apply your home-manager configuration.

Now it's simpler:
1. Edit the `keymaps.json` file.
2. Test it works.
3. If you want it to become part of your deterministic configuration, copy what you've done into your nix configuration, converted into Nix.

Changing the AI model was a nightmare too, as you said.

Also changing icons or colors, or testing themes.

If at any point you want to reset your settings, all you have to do is delete the files (at least after #7351 is merged).

I understand complete determinism has its benefits, but it's not practical for Zed.
